File: tests/test_mat_to_arr.py
"""Test armadillo matrix to  numpy array functions."""
import numpy as np

import test_carma as carma

###############################################################################
#                                 MAT
###############################################################################


def test_mat_to_arr():
    """Test mat_to_arr."""
    arr = carma.mat_to_arr(False)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_mat_to_arr_copy():
    """Test mat_to_arr."""
    arr = carma.mat_to_arr(True)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_mat_to_arr_plus_one():
    """Test mat_to_arr with c++ addition."""
    sample = np.asarray(
        np.random.normal(size=(20, 2)),
        dtype=np.float64,
        order='F'
    )
    mat = carma.mat_to_arr_plus_one(sample, False)
    assert np.allclose(mat, sample + 1)


def test_mat_to_arr_plus_one_copy():
    """Test mat_to_arr with c++ addition."""
    sample = np.asarray(
        np.random.normal(size=(20, 2)),
        dtype=np.float64,
        order='F'
    )
    mat = carma.mat_to_arr_plus_one(sample, True)
    assert np.allclose(mat, sample + 1)


def test_to_numpy_mat():
    arr = carma.to_numpy_mat(False)
    assert arr.ndim == 2
    assert arr.shape[0] == 4
    assert arr.shape[1] == 5
    assert arr.flags['WRITEABLE'] == True

    arr = carma.to_numpy_mat(True)
    assert arr.ndim == 2
    assert arr.shape[0] == 4
    assert arr.shape[1] == 5
    assert arr.flags['WRITEABLE'] == True


def test_to_numpy_view_mat():
    arr = carma.to_numpy_view_mat()
    assert arr.ndim == 2
    assert arr.shape[0] == 4
    assert arr.shape[1] == 5
    assert arr.flags['WRITEABLE'] == False


###############################################################################
#                                 ROW
###############################################################################


def test_row_to_arr():
    """Test row_to_arr."""
    arr = carma.row_to_arr(False)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_row_to_arr_copy():
    """Test row_to_arr."""
    arr = carma.row_to_arr(True)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_row_to_arr_plus_one():
    """Test row_to_arr with c++ addition."""
    sample = np.asarray(
        np.random.normal(size=(20)),
        dtype=np.float64,
        order='F'
    )
    mat = carma.row_to_arr_plus_one(sample, False)
    assert np.allclose(mat, sample + 1)


def test_row_to_arr_plus_one_copy():
    """Test row_to_arr with c++ addition."""
    sample = np.asarray(
        np.random.normal(size=(20)),
        dtype=np.float64,
        order='F'
    )
    mat = carma.row_to_arr_plus_one(sample, True)
    assert np.allclose(mat, sample + 1)


def test_to_numpy_row():
    arr = carma.to_numpy_row(False)
    assert arr.ndim == 2
    assert arr.shape[0] == 1
    assert arr.shape[1] == 20
    assert arr.flags['WRITEABLE'] == True

    arr = carma.to_numpy_row(True)
    assert arr.ndim == 2
    assert arr.shape[0] == 1
    assert arr.shape[1] == 20
    assert arr.flags['WRITEABLE'] == True


def test_to_numpy_view_row():
    arr = carma.to_numpy_view_row()
    assert arr.ndim == 2
    assert arr.shape[0] == 1
    assert arr.shape[1] == 20
    assert arr.flags['WRITEABLE'] == False


###############################################################################
#                                 COL
###############################################################################


def test_col_to_arr():
    """Test col_to_arr."""
    arr = carma.col_to_arr(False)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_col_to_arr_copy():
    """Test col_to_arr."""
    arr = carma.col_to_arr(True)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_col_to_arr_plus_one():
    """Test col_to_arr with c++ addition."""
    sample = np.asarray(
        np.random.normal(size=(20, 1)),
        dtype=np.float64,
        order='F'
    )
    mat = carma.col_to_arr_plus_one(sample, False)
    assert np.allclose(mat, sample + 1)


def test_col_to_arr_plus_one_copy():
    """Test col_to_arr with c++ addition."""
    sample = np.asarray(
        np.random.normal(size=(20, 1)),
        dtype=np.float64,
        order='F'
    )
    mat = carma.col_to_arr_plus_one(sample, True)
    assert np.allclose(mat, sample + 1)


def test_to_numpy_col():
    arr = carma.to_numpy_col(False)
    assert arr.shape[0] == 20
    assert arr.shape[1] == 1
    assert arr.flags['WRITEABLE'] == True

    arr = carma.to_numpy_col(True)
    assert arr.shape[0] == 20
    assert arr.shape[1] == 1
    assert arr.flags['WRITEABLE'] == True


def test_to_numpy_view_col():
    arr = carma.to_numpy_view_col()
    assert arr.ndim == 2
    assert arr.shape[0] == 20
    assert arr.shape[1] == 1
    assert arr.flags['WRITEABLE'] == False


###############################################################################
#                                 CUBE
###############################################################################


def test_cube_to_arr():
    """Test cube_to_arr."""
    arr = carma.cube_to_arr(False)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-16).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_cube_to_arr_return():
    """Test cube_to_arr."""
    arr = carma.cube_to_arr(True)

    nan_cnt = np.isnan(arr).sum()
    assert nan_cnt == 0, 'Conversion introduced NaNs'

    zero_cnt = (arr == 0.0).sum()
    assert zero_cnt == 0, 'Conversion introduced zeros'

    inf_cnt = (arr == np.inf).sum()
    assert inf_cnt == 0, 'Conversion introduced inf values'

    near_null = (np.abs(arr) <= 1e-6).sum()
    assert near_null == 0, 'Conversion introduced near zero values'


def test_cube_to_arr_plus_one():
    """Test cube_to_arr with c++ addition."""
    or_sample = np.random.normal(size=(20, 3, 2))
    sample = np.asarray(
        or_sample,
        dtype=np.float64,
        order='F'
    )
    mat = carma.cube_to_arr_plus_one(sample, False)
    assert np.allclose(mat, 1 + sample)


def test_cube_to_arr_plus_one_copy():
    """Test cube_to_arr with c++ addition."""
    or_sample = np.random.normal(size=(20, 3, 2))
    sample = np.asarray(
        or_sample,
        dtype=np.float64,
        order='F'
    )
    mat = carma.cube_to_arr_plus_one(sample, True)
    assert np.allclose(mat, 1 + sample)


def test_to_numpy_cube():
    arr = carma.to_numpy_cube(False)
    assert arr.ndim == 3
    assert arr.shape[0] == 100
    assert arr.shape[1] == 2
    assert arr.shape[2] == 4
    assert arr.flags['WRITEABLE'] == True

    arr = carma.to_numpy_cube(True)
    assert arr.ndim == 3
    assert arr.shape[0] == 100
    assert arr.shape[1] == 2
    assert arr.shape[2] == 4
    assert arr.flags['WRITEABLE'] == True


def test_to_numpy_view_cube():
    arr = carma.to_numpy_view_cube()
    assert arr.ndim == 3
    assert arr.shape[0] == 100
    assert arr.shape[1] == 2
    assert arr.shape[2] == 4
    assert arr.flags['WRITEABLE'] == False

File: tests/src/test_mat_to_arr.cpp
#include "test_mat_to_arr.h"

namespace carma {
namespace tests {

// ------------------------------ Mat -----------------------------------------
py::array_t<double> test_mat_to_arr(bool copy) {
    arma::mat mat = arma::mat(4, 5, arma::fill::randu);
    return mat_to_arr<double>(mat, copy);
} /* test_mat_to_arr */

py::array_t<double> test_to_numpy_mat(bool copy) {
    arma::mat mat = arma::mat(4, 5, arma::fill::randu);
    return to_numpy<arma::mat>(mat, copy);
} /* test_to_numpy_mat */

py::array_t<double> test_mat_to_arr_plus_one(const py::array_t<double>& arr, bool copy) {
    arma::Mat<double> ones = arma::ones(arr.shape(0), arr.shape(1));
    arma::Mat<double> mat = carma::arr_to_mat<double>(arr);
    arma::Mat<double> out = ones + mat;
    return mat_to_arr<double>(out, copy);
} /* test_mat_to_arr_plus_one */

py::array_t<double> test_to_numpy_view_mat() {
    const arma::mat mat = arma::mat(4, 5, arma::fill::randu);
    return to_numpy_view<arma::mat>(mat);
} /* test_to_numpy_mat */

// ------------------------------ Row -----------------------------------------
py::array_t<double> test_row_to_arr(bool copy) {
    arma::Row<double> mat = arma::Row<double>(20, arma::fill::randu);
    return row_to_arr<double>(mat, copy);
} /* test_row_to_arr */

py::array_t<double> test_to_numpy_row(bool copy) {
    arma::Row<double> mat = arma::Row<double>(20, arma::fill::randu);
    return to_numpy<arma::Row<double>>(mat, copy);
} /* test_to_numpy_row */

py::array_t<double> test_row_to_arr_plus_one(const py::array_t<double>& arr, bool copy) {
    arma::Row<double> ones = arma::Row<double>(arr.size(), arma::fill::ones);
    arma::Row<double> mat = carma::arr_to_row<double>(arr);
    arma::Row<double> out = ones + mat;
    return row_to_arr<double>(out, copy);
} /* test_row_to_arr_plus_one */

py::array_t<double> test_to_numpy_view_row() {
    const arma::Row<double> mat = arma::Row<double>(20, arma::fill::randu);
    return to_numpy_view<arma::Row<double>>(mat);
} /* test_to_numpy_row */

// ------------------------------ Col -----------------------------------------
py::array_t<double> test_col_to_arr(bool copy) {
    arma::Col<double> mat = arma::Col<double>(100, arma::fill::randu);
    return col_to_arr<double>(mat, copy);
} /* test_col_to_arr */

py::array_t<double> test_to_numpy_col(bool copy) {
    arma::Col<double> mat = arma::Col<double>(20, arma::fill::randu);
    return to_numpy<arma::Col<double>>(mat, copy);
} /* test_to_numpy_col */

py::array_t<double> test_col_to_arr_plus_one(const py::array_t<double>& arr, bool copy) {
    arma::Col<double> ones = arma::ones(arr.size());
    arma::Col<double> mat = carma::arr_to_col<double>(arr);
    arma::Col<double> out = ones + mat;
    return col_to_arr<double>(out, copy);
} /* test_col_to_arr_plus_one */

py::array_t<double> test_to_numpy_view_col() {
    const arma::Col<double> mat = arma::Col<double>(20, arma::fill::randu);
    return to_numpy_view<arma::Col<double>>(mat);
} /* test_to_numpy_col */

// ------------------------------ Cube ----------------------------------------
py::array_t<double> test_cube_to_arr(bool copy) {
    arma::Cube<double> mat = arma::Cube<double>(100, 2, 4, arma::fill::randu);
    return cube_to_arr<double>(mat, copy);
} /* test_cube_to_arr */

py::array_t<double> test_to_numpy_cube(bool copy) {
    arma::Cube<double> mat = arma::Cube<double>(100, 2, 4, arma::fill::randu);
    return to_numpy<arma::Cube<double>>(mat, copy);
} /* test_to_numpy_cube */

py::array_t<double> test_cube_to_arr_plus_one(const py::array_t<double>& arr, bool copy) {
    arma::Cube<double> ones = arma::ones(arr.shape(0), arr.shape(1), arr.shape(2));
    arma::Cube<double> mat = carma::arr_to_cube<double>(arr);
    arma::Cube<double> out = ones + mat;
    return cube_to_arr<double>(out, copy);
} /* test_mat_to_arr_plus_one */

py::array_t<double> test_to_numpy_view_cube() {
    const arma::Cube<double> mat = arma::Cube<double>(100, 2, 4, arma::fill::randu);
    return to_numpy_view<arma::Cube<double>>(mat);
} /* test_to_numpy_cube */

}  // namespace tests
}  // namespace carma

void bind_test_mat_to_arr(py::module& m) {
    m.def("mat_to_arr", &carma::tests::test_mat_to_arr, "Test mat_to_arr");
}

void bind_test_row_to_arr(py::module& m) {
    m.def("row_to_arr", &carma::tests::test_row_to_arr, "Test mat_to_arr");
}

void bind_test_col_to_arr(py::module& m) {
    m.def("col_to_arr", &carma::tests::test_col_to_arr, "Test col_to_arr");
}

void bind_test_cube_to_arr(py::module& m) {
    m.def("cube_to_arr", &carma::tests::test_cube_to_arr, "Test cube_to_arr");
}

void bind_test_to_numpy_mat(py::module& m) {
    m.def("to_numpy_mat", &carma::tests::test_to_numpy_mat, "Test to_numpy");
}

void bind_test_to_numpy_row(py::module& m) {
    m.def("to_numpy_row", &carma::tests::test_to_numpy_row, "Test to_numpy");
}

void bind_test_to_numpy_col(py::module& m) {
    m.def("to_numpy_col", &carma::tests::test_to_numpy_col, "Test to_numpy");
}

void bind_test_to_numpy_cube(py::module& m) {
    m.def("to_numpy_cube", &carma::tests::test_to_numpy_cube, "Test to_numpy");
}

void bind_test_to_numpy_view_mat(py::module& m) {
    m.def("to_numpy_view_mat", &carma::tests::test_to_numpy_view_mat, "Test to_numpy");
}

void bind_test_to_numpy_view_row(py::module& m) {
    m.def("to_numpy_view_row", &carma::tests::test_to_numpy_view_row, "Test to_numpy");
}

void bind_test_to_numpy_view_col(py::module& m) {
    m.def("to_numpy_view_col", &carma::tests::test_to_numpy_view_col, "Test to_numpy");
}

void bind_test_to_numpy_view_cube(py::module& m) {
    m.def("to_numpy_view_cube", &carma::tests::test_to_numpy_view_cube, "Test to_numpy");
}

void bind_test_mat_to_arr_plus_one(py::module& m) {
    m.def("mat_to_arr_plus_one", &carma::tests::test_mat_to_arr_plus_one, "Test mat_to_arr");
}

void bind_test_row_to_arr_plus_one(py::module& m) {
    m.def("row_to_arr_plus_one", &carma::tests::test_row_to_arr_plus_one, "Test row_to_arr");
}

void bind_test_col_to_arr_plus_one(py::module& m) {
    m.def("col_to_arr_plus_one", &carma::tests::test_col_to_arr_plus_one, "Test col_to_arr");
}

void bind_test_cube_to_arr_plus_one(py::module& m) {
    m.def("cube_to_arr_plus_one", &carma::tests::test_cube_to_arr_plus_one, "Test cube_to_arr");
}
